### PR TITLE
Fix a bunch of pretty-printer bugs

### LINKED ISSUE #2650: some additional pretty printer round-trip bugs
```unison:hide
    broken =
        addNumbers: 'Nat
        addNumbers = 'let
          use Nat +
          y = 12
          13 + y
        !addNumbers    
    ```
    
    ``` ucm
    .> add
    .> edit broken
    .> load scratch.u
    .> update
    .> undo
    ```
    
    
    ```unison:hide
    
    tvarmodify tvar fun = ()
    
    broken tvar = 
      '(tvarmodify tvar (cases
         Some _ -> "oh boy isn't this a very very very very very very very long string?"
         None -> ""))
         
    ```
    
    ``` ucm
    .> add
    .> edit broken
    .> load scratch.u
    .> update
    .> undo
    ```
    
    ```unison:hide
    
    broken = cases 
      Some loooooooooooooooooooooooooooooooooooooooooooooooooooooooong | loooooooooooooooooooooooooooooooooooooooooooooooooooooooong == 1 -> ()
    
    ```
    
    ``` ucm
    .> add
    .> edit broken
    .> load scratch.u
    .> update
    .> undo
    ```

### PR DESCRIPTION
**Pull request by runarorama**
This fixes #2650 which covers a bunch of pretty-printer bugs.

### CODE REVIEW

File: parser-typechecker/src/Unison/TermPrinter.hs [hunk 1]
```diff
@@ -1097,65 +1106,104 @@ calcImports im tm = (im', render $ getUses result)
...
-    If' _ t f -> [t, f]
-    LetBlock bs _ -> concat $ map doLet bs
-    Match' scrute branches ->
-      if isDestructuringBind scrute branches then [scrute]
-      else concat $ map doCase branches
-    _ -> []
+  Handle' handler body -> [handler, body]
+  If' _ t f -> [t, f]
+  LetBlock bs e -> concatMap doLet bs ++ handleDelay e
+  Delay' b@(Lets' _ _) -> [b]
```
Comment (runarorama) [inline parser-typechecker/src/Unison/TermPrinter.hs:1193]: Added this special case as `'let` introduces a syntactic block that is (syntactically) an immediate child of the surrounding block.

File: parser-typechecker/src/Unison/TermPrinter.hs [hunk 2]
```diff
@@ -1097,65 +1106,104 @@ calcImports im tm = (im', render $ getUses result)
...
-    Handle' handler body -> [handler, body]
-    If' _ t f -> [t, f]
-    LetBlock bs _ -> concat $ map doLet bs
-    Match' scrute branches ->
-      if isDestructuringBind scrute branches then [scrute]
-      else concat $ map doCase branches
-    _ -> []
+  Handle' handler body -> [handler, body]
+  If' _ t f -> [t, f]
+  LetBlock bs e -> concatMap doLet bs ++ handleDelay e
```
Comment (runarorama) [inline parser-typechecker/src/Unison/TermPrinter.hs:1192]: Added the case where `e` (the expression at the end of the `let` block) is a `'let` block.

File: parser-typechecker/src/Unison/TermPrinter.hs [hunk 3]
```diff
@@ -248,7 +248,13 @@ pretty0
     Delay' x ->
       paren (p >= 11 || isBlock x && p >= 3) $
         fmt S.DelayForceChar (l "'")
-        <> pretty0 n (ac (if isBlock x then 0 else 10) Normal im doc) x
+        <> (case x of
+              Lets' _ _ -> id
+              -- Add indentation below if we're opening parens with '(
+              -- This is in case the contents are a long function application
+              -- in which case the arguments should be indented.
+              _ -> PP.indentAfterNewline "  ")
```
Comment (runarorama) [inline parser-typechecker/src/Unison/TermPrinter.hs:256]: Avoids a virtual semicolon when line-breaking long function applications after a `'(`

File: parser-typechecker/src/Unison/TermPrinter.hs [hunk 4]
```diff
@@ -571,7 +578,9 @@ printCase env im doc ms0 = PP.lines $ map each gridArrowsAligned where
   printGuard (Just (ABT.AbsN' _ g)) =
     -- strip off any Abs-chain around the guard, guard variables are rendered
     -- like any other variable, ex: case Foo x y | x < y -> ...
-    PP.group $ PP.spaced [(fmt S.DelimiterChar " |"), pretty0 env (ac 2 Normal im doc) g]
+    PP.group $
+      PP.indentAfterNewline "  " $
```
Comment (runarorama) [inline parser-typechecker/src/Unison/TermPrinter.hs]: Adds indentation in the case that the guard doesn't fit on the same line as the pattern.

**Review by pchiusano — APPROVED** [review]
Yay!

**Comment by ceedubs** [comment]
@runarorama we gave these a try and they are working great! They have eliminated at least a dozen roundtrip bugs in our Unison Cloud codebase.

**Pull request merged**